models/DataObject/Concrete/Dao/InheritanceHelper.php

```diff
@@ -418,15 +429,53 @@ protected function filterResultByLanguage($result, $language, $column)
     protected function buildTree($currentParentId, $fields = '', $parentIdGroups = null, $params = [])
     {
         $objects = [];
+        $storeTable = $this->storetable;
+        $idfield = $this->idField;
 
         if (!$parentIdGroups) {
             $object = DataObject::getById($currentParentId);
             if (isset($params['language'])) {
-                $query = "SELECT a.language as language, b.o_id AS id $fields, b.o_classId AS classId, b.o_parentId AS parentId FROM objects b LEFT JOIN " . $this->storetable . ' a ON b.o_id = a.' . $this->idField . ' WHERE o_path LIKE ' . $this->db->quote($this->db->escapeLike($object->getRealFullPath()) . '/%')
-                    . ' HAVING `language` = "' . $params['language'] . '" OR ISNULL(`language`)'
-                    . ' ORDER BY LENGTH(o_path) ASC';
+
+                $language = $params['language'];
+
+                $query = "
+                WITH RECURSIVE cte(id, classId, parentId, o_path) as (
+                    SELECT c.o_id AS id, c.o_classId AS classId, c.o_parentId AS parentId, c.o_path AS o_path
+                    FROM objects c
+                    WHERE c.o_parentId = $currentParentId
+                    UNION ALL
+                    SELECT p.o_id AS id, p.o_classId AS classId, p.o_parentId AS parentId, p.o_path AS o_path
+                    FROM objects p
+                    INNER JOIN cte on (p.o_parentId = cte.id)
+                ) SELECT l.language AS `language`,
+                         x.id AS id,
+                         x.classId AS classId,
+                         x.parentId AS parentId
+                         $fields
+                    FROM cte x
+                    LEFT JOIN $storeTable l ON x.id = l.$idfield
+                   WHERE COALESCE(`language`, " . $this->db->quote($language) . ") = " . $this->db->quote($language) .
```

Comment:
Hi @aclDeveloper can you explain with an example how the coalesce is equivalent to the having clause as it was before?

Reply:
So the original query utilizes a `HAVING` clause to implicitly group languages together. However there is no `GROUP BY` clause present, so it just behaves as the `WHERE` would, as you can see in the following 2 examples:

![image](https://user-images.githubusercontent.com/20439617/115001903-64543f00-9ea4-11eb-9309-878f3a5d972a.png)

![image](https://user-images.githubusercontent.com/20439617/115001989-81890d80-9ea4-11eb-9982-4f053a262249.png)

With the recursive query we more or less have the same behavior. I propose to put the language filter in the `WHERE` clause because there's no `GROUP BY` and so its easier to read for people who are not extra-fond in SQL. 

![image](https://user-images.githubusercontent.com/20439617/115002839-7b476100-9ea5-11eb-9e2b-0212e2fe7ff0.png)

The use of COALESCE is used as a short-hand here so that we don't have to incorporate the `OR ISNULL(language)` part. If we replace NULL-Values with the expected value, we can get the same result and save on one more condition on an already complicated query.